File: packages/android/core-domain/src/main/kotlin/app/notes/domain/Storage.kt
package app.notes.domain

import java.io.InputStream

/**
 * The cross-cutting storage port. Concrete adapters live in
 * `storage-android`. Consumers depend on this interface only.
 */
interface StorageProvider {
    val id: StorageProviderId
    suspend fun upload(file: FileInput, folderHint: String? = null): RemoteFile
    suspend fun getDownloadUrl(externalId: String, ttlSeconds: Int = 300): String
    suspend fun delete(externalId: String)
    /** Create / find the per-app folder, returning its provider-side id. Cached. */
    suspend fun ensureAppFolder(): String
}

data class RemoteFile(
    val externalId: String,
    val name: String,
    val mimeType: String,
    val sizeBytes: Long,
) {
    init {
        require(externalId.isNotBlank()) { "RemoteFile.externalId cannot be empty" }
        require(name.isNotBlank()) { "RemoteFile.name cannot be empty" }
        require(sizeBytes >= 0) { "RemoteFile.sizeBytes must be >= 0" }
    }
}

/**
 * Either [bytes] (for small files) or [stream] (for resumable uploads) must
 * be present. Adapters pick based on [sizeBytes] vs
 * [SIMPLE_UPLOAD_THRESHOLD_BYTES].
 */
data class FileInput(
    val name: String,
    val mimeType: String,
    val sizeBytes: Long,
    val bytes: ByteArray? = null,
    val stream: (() -> InputStream)? = null,
) {
    init {
        require(name.isNotBlank()) { "FileInput.name cannot be empty" }
        require(mimeType.isNotBlank()) { "FileInput.mimeType cannot be empty" }
        require(sizeBytes >= 0) { "FileInput.sizeBytes must be >= 0" }
        require(bytes != null || stream != null) {
            "FileInput needs either bytes or stream"
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is FileInput) return false
        return name == other.name &&
            mimeType == other.mimeType &&
            sizeBytes == other.sizeBytes &&
            bytes contentEqualsOrBothNull other.bytes &&
            (stream == null) == (other.stream == null)
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + mimeType.hashCode()
        result = 31 * result + sizeBytes.hashCode()
        result = 31 * result + (bytes?.contentHashCode() ?: 0)
        return result
    }
}

private infix fun ByteArray?.contentEqualsOrBothNull(other: ByteArray?): Boolean =
    if (this == null && other == null) true
    else if (this == null || other == null) false
    else this.contentEquals(other)

const val SIMPLE_UPLOAD_THRESHOLD_BYTES: Long = 5L * 1024L * 1024L

sealed class StorageError(val message: String) {
    class Unauthorized(message: String) : StorageError(message)
    class QuotaExceeded(message: String) : StorageError(message)
    class NotFound(message: String) : StorageError(message)
    class TransportError(message: String) : StorageError(message)
    class UpstreamError(val status: Int, message: String) : StorageError(message)
}

class StorageException(val error: StorageError) :
    RuntimeException("[${error::class.simpleName}] ${error.message}")

File: packages/domain/src/storage.ts
import type { StorageProviderId } from "./enums.js";

/**
 * The cross-cutting storage port. Concrete adapters
 * (GoogleDriveAdapter in V1) live in `@notes-app/storage`. Consumers
 * (M07 Attachment Pipeline, M06 editor's `insertAttachment`) depend on
 * this interface only — they never know whether the bytes land in
 * Drive, OneDrive, or S3.
 */
export interface StorageProvider {
  readonly id: StorageProviderId;
  /**
   * Upload [file] to the provider, returning the persistent reference
   * the rest of the app stores in `attachments_refs.external_id`.
   * `folderHint` is advisory — adapters may use it (e.g. as a Drive
   * appProperties tag) or ignore it.
   */
  upload(file: FileInput, folderHint?: string): Promise<RemoteFile>;
  /** Short-lived (≤ ttlSeconds, default 300) URL the client can fetch from. */
  getDownloadUrl(externalId: string, ttlSeconds?: number): Promise<string>;
  delete(externalId: string): Promise<void>;
  /** Create / find the per-app folder, returning its provider-side id. Cached. */
  ensureAppFolder(): Promise<string>;
}

export interface RemoteFile {
  readonly externalId: string;
  readonly name: string;
  readonly mimeType: string;
  readonly sizeBytes: number;
}

/**
 * Input contract for an upload. Adapters pick between [bytes] and [stream]
 * based on size: anything under [SIMPLE_UPLOAD_THRESHOLD_BYTES] goes via
 * the small-file path; larger files use resumable upload and pull from
 * [stream] lazily so we don't materialise the whole file in memory.
 */
export interface FileInput {
  readonly bytes?: Uint8Array;
  readonly stream?: () => AsyncIterable<Uint8Array>;
  readonly name: string;
  readonly mimeType: string;
  readonly sizeBytes: number;
}

export const SIMPLE_UPLOAD_THRESHOLD_BYTES = 5 * 1024 * 1024; // 5 MB

/** Provider-agnostic storage failure modes. */
export type StorageError =
  | { kind: "Unauthorized"; message: string }
  | { kind: "QuotaExceeded"; message: string }
  | { kind: "NotFound"; message: string }
  | { kind: "TransportError"; message: string }
  | { kind: "UpstreamError"; status: number; message: string };

export const StorageError = {
  unauthorized: (message: string): StorageError => ({ kind: "Unauthorized", message }),
  quotaExceeded: (message: string): StorageError => ({ kind: "QuotaExceeded", message }),
  notFound: (message: string): StorageError => ({ kind: "NotFound", message }),
  transport: (message: string): StorageError => ({ kind: "TransportError", message }),
  upstream: (status: number, message: string): StorageError => ({
    kind: "UpstreamError",
    status,
    message,
  }),
} as const;

export class StorageException extends Error {
  readonly code = "ERR_STORAGE" as const;
  constructor(public readonly error: StorageError) {
    super(`[${error.kind}] ${error.message}`);
  }
}
